File: libs/core/src/symbols/function_symbol.cpp
/**
 * @file function_symbol.cpp
 * @author Francesco Corso
 * @date 8/8/2026
 * @brief
 */
module prism.core:symbols.function_symbol.impl;

import :symbols.function_symbol;
import :symbols.type_symbol;
import :symbols.assembly_symbol;
import :symbols.parameter_symbol;

namespace prism
{

    bool FunctionSymbol::returns_void() const
    {
        return return_type().is_void();
    }

    void FunctionSymbol::write_display_string(TextWriter &writer) const
    {
        auto containing = containing_symbol();
        if (containing.has_value() && !containing->is<AssemblySymbol>())
        {
            containing->write_display_string(writer);
            writer.write("::");
        }

        writer.write(name().as_string_view());
        writer.write('(');
        for (auto [i, param] : parameters() | std::views::enumerate)
        {
            if (i > 0)
                writer.write(", ");

            param->write_display_string(writer);
        }
        writer.write(')');

        if (!returns_void())
        {
            writer.write(": ");
            return_type().write_display_string(writer);
        }
    }
} // namespace prism

File: src/Prism.Core/Symbols/FunctionSymbol.cs
using System.Collections.Immutable;
using Prism.Core.Abi;
using ZLinq;

namespace Prism.Core.Symbols;

public closed class FunctionSymbol : Symbol
{
    private protected FunctionSymbol(string name, Symbol? containingSymbol)
        : base(name, containingSymbol) { }

    public abstract TypeSymbol ReturnType { get; }

    public abstract ImmutableArray<ParameterSymbol> Parameters { get; }

    public bool ReturnsVoid => ReturnType.IsVoid;

    internal abstract AbiKind AbiKind { get; }

    internal ImmutableArray<TypeSymbol> ParameterTypes
    {
        get
        {
            if (!field.IsDefault)
                return field;

            ImmutableInterlocked.InterlockedInitialize(
                ref field,
                [.. Parameters.Select(p => p.Type)]
            );
            return field;
        }
    }

    public abstract bool IsExtern { get; }

    public sealed override void WriteDisplayString(TextWriter writer)
    {
        if (
            ContainingSymbol
            is not null
                and not AssemblySymbol
                and NamespaceSymbol { IsGlobal: false }
        )
        {
            ContainingSymbol.WriteDisplayString(writer);
            writer.Write("::");
        }

        writer.Write(Name);
        writer.Write('(');
        foreach (var (i, param) in Parameters.AsValueEnumerable().Index())
        {
            if (i > 0)
                writer.Write(", ");

            param.WriteDisplayString(writer);
        }
        writer.Write(')');
        if (ReturnsVoid)
            return;

        writer.Write(": ");
        ReturnType.WriteDisplayString(writer);
    }
}
